BedrockCore.cpp

```diff
@@ -64,6 +64,98 @@ bool BedrockCore::isTimedOut(unique_ptr<BedrockCommand>& command) {
+        command->prePeekCount++;
+
+        _db.startTiming(timeout);
+
+        try {
+            if (!_db.beginTransaction(SQLite::TRANSACTION_TYPE::SHARED)) {
+                STHROW("501 Failed to begin shared prePeek transaction");
+            }
+
+            // Make sure no writes happen while in prePeek command
+            _db.setQueryOnly(true);
+
+            // prePeek.
+            command->reset(BedrockCommand::STAGE::PREPEEK);
+            bool completed = command->prePeek(_db);
+            SDEBUG("Plugin '" << command->getName() << "' prePeeked command '" << request.methodLine << "'");
+
+            if (command->httpsRequests.size()) {
+                STHROW("405 https requests cannot be made in prePeek");
+            }
+
+            if (!completed) {
+                SDEBUG("Command '" << request.methodLine << "' not finished in prePeek, re-queuing.");
+                _db.rollback();
+                _db.resetTiming();
+                return RESULT::SHOULD_PEEK;
+            }
+
+        } catch (const SQLite::timeout_error& e) {
+            // Some plugins want to alert timeout errors themselves, and make them silent on bedrock.
+            if (!command->shouldSuppressTimeoutWarnings()) {
+                SALERT("Command " << command->request.methodLine << " timed out after " << e.time()/1000 << "ms.");
+            }
+            STHROW("555 Timeout prePeeking command");
+        }
+
+        // If no response was set, assume 200 OK
+        if (response.methodLine == "") {
+            response.methodLine = "200 OK";
+        }
+
+        // Add the commitCount header to the response.
+        response["commitCount"] = to_string(_db.getCommitCount());
+
+        // Success. If a command has set "content", encode it in the response.
+        SINFO("Responding '" << response.methodLine << "' to read-only '" << request.methodLine << "'.");
+        if (!content.empty()) {
+            // Make sure we're not overwriting anything different.
+            string newContent = SComposeJSONObject(content);
+            if (response.content != newContent) {
+                if (!response.content.empty()) {
+                    SWARN("Replacing existing response content in " << request.methodLine);
+                }
+                response.content = newContent;
+            }
+        }
+    } catch (const SException& e) {
+        _handleCommandException(command, e);
+    } catch (...) {
+        SALERT("Unhandled exception typename: " << SGetCurrentExceptionName() << ", command: " << request.methodLine);
+        command->response.methodLine = "500 Unhandled Exception";
+    }
+
+    // Unless an exception handler set this to something different, the command is complete.
+    command->complete = returnValue == RESULT::COMPLETE;
+
+    // Back out of the current transaction, it doesn't need to do anything.
+    _db.rollback();
+    _db.resetTiming();
+
+    // Reset, we can write now.
+    _db.setQueryOnly(false);
```

Comment:
We want to do this regardless, even if we're not complete, correct?

I think supporting `COMPLETE` here makes this a lot more complex. Commands should almost never be complete here. If we remove the ability to return `COMPLETE` here, we can assume we'll always run `peek` after this, but this function will be a lot simpler.

Reply:
I'm fine with removing the ability for a command to complete in the `prePeek`.

Reply:
So, in that case `prePeek` will not return a `bool`, but will rather be a `void`??